burstphoto/align.swift

```diff
@@ -399,69 +1080,29 @@ func color_difference(_ texture1: MTLTexture, _ texture2: MTLTexture, _ mosaic_p
-    let state = average_y_state
-    command_encoder.setComputePipelineState(state)
-    let threads_per_grid = MTLSize(width: in_texture.width, height: 1, depth: 1)
-    let max_threads_per_thread_group = state.maxTotalThreadsPerThreadgroup
-    let threads_per_thread_group = MTLSize(width: max_threads_per_thread_group, height: 1, depth: 1)
-    command_encoder.setTexture(in_texture, index: 0)
-    command_encoder.setTexture(avg_y, index: 1)
-    command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
-    
-    // average the generated 1d texture along the x-axis
-    let state2 = average_x_state
-    command_encoder.setComputePipelineState(state2)
-    let avg_buffer = device.makeBuffer(length: MemoryLayout<Float32>.size, options: .storageModeShared)!
-    command_encoder.setTexture(avg_y, index: 0)
-    command_encoder.setBuffer(avg_buffer, offset: 0, index: 0)
-    command_encoder.setBytes([Int32(in_texture.width)], length: MemoryLayout<Int32>.stride, index: 1)
-    command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
-    command_encoder.endEncoding()
-    command_buffer.commit()
-    
-    // return the average of all pixels in the input array
-    return avg_buffer
-}
-
-
-func estimate_color_noise(_ texure: MTLTexture, _ texture_blurred: MTLTexture, _ mosaic_pettern_width: Int) -> MTLBuffer {
+func estimate_color_noise(_ texture: MTLTexture, _ texture_blurred: MTLTexture, _ mosaic_pattern_width: Int) -> MTLBuffer {
     
     // compute the color difference of each mosaic superpixel between the original and the blurred texture
-    let texture_diff = color_difference(texure, texture_blurred, mosaic_pettern_width)
+    let texture_diff = color_difference(texture, texture_blurred, mosaic_pattern_width)
     
     // compute the average of the difference between the original and the blurred texture
-    let mean_diff = texture_mean(texture_diff)
+    let mean_diff = texture_mean(texture_diff, "r")
     
     return mean_diff
 }
 
 
-func robust_merge(_ ref_texture: MTLTexture, _ ref_texture_blurred: MTLTexture, _ comp_texture: MTLTexture, _ kernel_size: Int, _ robustness: Double, _ noise_sd: MTLBuffer, _ mosaic_pettern_width: Int) -> MTLTexture {
+func robust_merge(_ ref_texture: MTLTexture, _ ref_texture_blurred: MTLTexture, _ comp_texture: MTLTexture, _ kernel_size: Int, _ robustness: Double, _ noise_sd: MTLBuffer, _ mosaic_pattern_width: Int) -> MTLTexture {
     
-    // if robustness is 0, return simply the comparison texture
-    if robustness == 0 {return comp_texture}
-    
-    // blur comp texure
-    let comp_texture_blurred = blur_mosaic_texture(comp_texture, kernel_size, mosaic_pettern_width)
+    // blur comparison texture
+    var comp_texture_blurred = blur_mosaic_texture(comp_texture, kernel_size, mosaic_pattern_width)
+    comp_texture_blurred = blur_mosaic_texture(comp_texture_blurred, kernel_size, mosaic_pattern_width)
```

Comment:
Why does this it get blurred twice?

Reply:
Blurring twice increases the strength of blurring while still using blurring kernels with a relatively small support. I tested against a larger blurring kernel applied only once and there was no real benefit in calculation time.

Reply:
Was blurring only once not enough?

Reply:
In the spatial merging, the blurring is required to split the image into low frequency and high frequency components. It is difficult to answer, but I had the feeling that stronger blurring yields slightly better results (from a subjective point of view).

Reply:
Added the blur kernels for a kernel size of 16 today. As a consequence, only one call of the blur function is required.